zadanie3/ALS.cs

```diff
@@ -8,5 +13,180 @@ public class ALS
+        {
+            var listOfIndexes = new List<int>();
+            // ma przeszukac wiersz u i znalezc            
+            for (int j = 0; j < array.GetLength(1); j++) // skopiowany z Utility, nie rozumiem tego: j < array.GetLength(1)
+            {
+                if (array[u, j] != 0)
+                    listOfIndexes.Add(j);
+            }
+            return listOfIndexes;
+        }// sprawdza ktore indeksy sa rozne od 0
+
+        public List<int> FlatNoZeroOnColumn(int p, int[,] array) //  In [38]:    arg: ilosc uzytkownikow, array z ocenami. Zwraca array z indeksami != 0 dla wiersza = u. 
+        {
+            var listOfIndexes = new List<int>();
+            // ma przeszukac wiersz u i znalezc            
+            for (int j = 0; j < array.GetLength(0); j++) // skopiowany z Utility, nie rozumiem tego: j < array.GetLength(1)
+            {
+                if (array[j, p] != 0)
+                    listOfIndexes.Add(j);
+            }
+            return listOfIndexes;
+        }// sprawdza ktore indeksy sa rozne od 0
+
+
+        public float[,] TakeIndexValues(List<int> listOfIndexes, float[,] array, float d)
+        {
+
+            // czyli kolumny z macierzy P o indeksach w _I_u_)
+            var arrayIndexValues = new float[(int)d, listOfIndexes.Count];
+            // np listOfIndexes to [4,6,7] to wynikiem ma byc dwuwymiarowa tablica wartosci o indexach [ [ [0,4],[0,6],[0,7] ] , [ [1,4], [1,6], [1,7] ] , ... , [ [n,4], [n,6], [n,7] ] ],
+            //a wartosci pod tymi indeksami wziete z array, a array to lista random_product
+            for (int i = 0; i < d; i++)
+            {
+                var j = 0;
+                foreach (var index in listOfIndexes)
+                {
+                    arrayIndexValues[i, j++] = array[i, index];
+                }
+            }
+            return arrayIndexValues;
+
+        }
+
+
+       
+
+        public float[,] CreateEye(float size) // stworz macierz jednostowa o zadanym rozmiarze
+        {
+            var eye = new float[,] { };
+            for (int i = 0; i < size; i++)
+            {
+                for (int j = 0; j < size; i++)
+                {
+                    if (i == j) eye[i, j] = 1;
+                    else eye[i, j] = 0;
+                }
+            }
+            return eye;
+            // return macierz_jednostkowa
+        }
+
+
+        //TO DO
+        //A_u = np.matmul(P_I_u, P_I_u_T) + reg* E 
+        // do tego w sumie potrzebna metoda transponowania, a reszte w sumie mamy. mozna to eventualnie zamknac w metode dokonujaca tego mnozenia kilku elementow.
+
+
+        /*Następnie liczymy _V_u_, czyli musimy dodać do siebie:
+
+        kolumnę 4 z macierzy P pomnożoną przez 4 (ocena)
+        kolumnę 6 z macierzy P pomnożoną przez 5 (ocena)
+        kolumnę 7 z macierzy P pomnożoną przez 4 (ocena)
+        */
+        public float[,] createCloneArray(float[,] matrix)
```

Comment:
Lepiej użyć wbudowaną metodę
https://www.tutorialspoint.com/Array-Copy-in-Chash